File: Java/src/test/java/minimal/CookingServiceTest.java
package minimal;

import static minimal.model.Name.named;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.util.Arrays;
import java.util.List;

import minimal.model.Ingredient;
import minimal.model.Name;
import minimal.model.NoSuchIngredientException;
import minimal.model.Recipe;

import org.junit.Test;

public class CookingServiceTest {

    private static final Name EGG = named("Egg");
    private static final Name CHEESE = named("Cheese");
    private static final Name SALT = named("Salt");
    private static final Name OMELETTE = named("Omelette");

    private CookingService cookingService = new CookingService();

    @Test
    public void testWhatDoWeHaveToEat() {
        List<Name> ingredients = cookingService.listIngredients();

        assertEquals(Arrays.asList(EGG, CHEESE, SALT), ingredients);
    }

    @Test
    public void testTakeOut() {
        Ingredient cheese = cookingService.takeOut(CHEESE);

        assertNotNull(cheese);
        assertTrue("opened", cheese.isOpen());
    }

    @Test
    public void testTakeOutCaseInsensitive() {
        Ingredient egg = cookingService.takeOut(named("egg"));

        assertNotNull(egg);
    }

    @Test(expected = NoSuchIngredientException.class)
    public void testTakeOutMissing() {
        cookingService.takeOut(named("Bacon"));
    }

    @Test
    public void testHowToCook() {
        Recipe omelette = cookingService.howToCook(OMELETTE);

        assertNotNull(omelette);
        assertEquals(Arrays.asList(EGG, CHEESE), omelette.listIngredients());
    }

    @Test
    public void testTakeOutAllFor() {
        Recipe omelette = cookingService.howToCook(OMELETTE);
        List<Ingredient> ingredients = cookingService.takeOutAllFor(omelette);

        assertEquals(Arrays.asList(Ingredient.of(EGG), Ingredient.of(CHEESE)), ingredients);
    }

    @Test
    public void testAddToMyCookbook() {
        Name somethingTasty = named("test recipe");
        cookingService.addToMyCookbook(new Recipe(somethingTasty, named("Egg")));

        Recipe recipe = cookingService.howToCook(somethingTasty);

        assertNotNull(recipe);
        assertEquals(Arrays.asList(EGG), recipe.listIngredients());
    }

    @Test
    public void testremoveFromCookbook() {
        cookingService.removeFromCookbook(OMELETTE);

        assertFalse("no Omelet", cookingService.getCookbook().tableOfContents().contains(OMELETTE));
    }

    @Test(expected = NoSuchIngredientException.class)
    public void testremoveFromCookbook2() {
        cookingService.removeFromCookbook(OMELETTE);
        cookingService.takeOut(OMELETTE);
    }
}

File: PHP/tests/CookingServiceTest.php
<?php

namespace Minimal;

use PHPUnit\Framework\TestCase;
use Minimal\Model\Name;
use Minimal\Model\Ingredient;
use Minimal\Model\Recipe;
use Minimal\Model\Cookbook;
use Minimal\Dao\Database;

class CookingServiceTest extends TestCase
{
    private CookingService $cookingService;

    private static Name $EGG;
    private static Name $CHEESE;
    private static Name $SALT;
    private static Name $OMELETTE;

    public function setUp(): void
    {
        $this->cookingService = new CookingService();

        self::$EGG = Name::named("Egg");
        self::$CHEESE = Name::named("Cheese");
        self::$SALT = Name::named("Salt");
        self::$OMELETTE = Name::named("Omelette");
    }

    public function testWhatDoWeHaveToEat()
    {
        $ingredients = $this->cookingService->listIngredients();

        $expectedIngredients = [self::$EGG, self::$CHEESE, self::$SALT];
        $this->assertEquals($expectedIngredients, $ingredients);
    }

    public function testTakeOut()
    {
        $cheese = $this->cookingService->takeOut(self::$CHEESE);

        $this->assertInstanceOf(Ingredient::class, $cheese);
        $this->assertTrue($cheese->isOpen());
    }

    public function testTakeOutCaseInsensitive()
    {
        $egg = $this->cookingService->takeOut(Name::named("egg"));

        $this->assertInstanceOf(Ingredient::class, $egg);
    }

    public function testTakeOutMissing()
    {
        $this->expectException(NoSuchIngredientException::class);
        $this->cookingService->takeOut(Name::named("Bacon"));
    }

    public function testHowToCook()
    {
        $omelette = $this->cookingService->howToCook(self::$OMELETTE);

        $this->assertInstanceOf(Recipe::class, $omelette);
        $expectedIngredients = [self::$EGG, self::$CHEESE];
        $this->assertEquals($expectedIngredients, $omelette->listIngredients());
    }

    public function testTakeOutAllFor()
    {
        $omelette = $this->cookingService->howToCook(self::$OMELETTE);
        $ingredients = $this->cookingService->takeOutAllFor($omelette);

        $expectedIngredients = [Ingredient::of(self::$EGG), Ingredient::of(self::$CHEESE)];
        $this->assertEquals($expectedIngredients, $ingredients);
    }

    public function testAddToMyCookbook()
    {
        $somethingTasty = Name::named("test recipe");
        $this->cookingService->addToMyCookbook(new Recipe($somethingTasty, self::$EGG));

        $recipe = $this->cookingService->howToCook($somethingTasty);

        $this->assertInstanceOf(Recipe::class, $recipe);
        $this->assertEquals([self::$EGG], $recipe->listIngredients());
    }

    public function testRemoveFromCookbook()
    {
        $this->cookingService->removeFromCookbook(self::$OMELETTE);

        $cookbook = $this->cookingService->getCookbook();
        $this->assertFalse(in_array(self::$OMELETTE, $cookbook->tableOfContents()));
    }

    public function testRemoveFromCookbookTwice()
    {
        $this->cookingService->removeFromCookbook(self::$OMELETTE);

        $this->expectException(NoSuchIngredientException::class);
        $this->cookingService->takeOut(self::$OMELETTE);
    }
}
